subsys/bluetooth/mesh/light_hsl_srv.c

```diff
@@ -0,0 +1,549 @@
+	{
+		BT_MESH_LIGHT_HSL_OP_GET,
+		BT_MESH_LIGHT_HSL_MSG_LEN_GET,
+		hsl_get_handle,
+	},
+	{
+		BT_MESH_LIGHT_HSL_OP_SET,
+		BT_MESH_LIGHT_HSL_MSG_MINLEN_SET,
+		hsl_set_handle,
+	},
+	{
+		BT_MESH_LIGHT_HSL_OP_SET_UNACK,
+		BT_MESH_LIGHT_HSL_MSG_MINLEN_SET,
+		hsl_set_unack_handle,
+	},
+	{
+		BT_MESH_LIGHT_HSL_OP_TARGET_GET,
+		BT_MESH_LIGHT_HSL_MSG_LEN_GET,
+		hsl_target_get_handle,
+	},
+	{
+		BT_MESH_LIGHT_HSL_OP_DEFAULT_GET,
+		BT_MESH_LIGHT_HSL_MSG_LEN_GET,
+		default_get_handle,
+	},
+	{
+		BT_MESH_LIGHT_HSL_OP_RANGE_GET,
+		BT_MESH_LIGHT_HSL_MSG_LEN_GET,
+		range_get_handle,
+	},
+	BT_MESH_MODEL_OP_END,
+};
+
+const struct bt_mesh_model_op _bt_mesh_light_hsl_setup_srv_op[] = {
+	{
+		BT_MESH_LIGHT_HSL_OP_DEFAULT_SET,
+		BT_MESH_LIGHT_HSL_MSG_LEN_DEFAULT,
+		default_set_handle,
+	},
+	{
+		BT_MESH_LIGHT_HSL_OP_DEFAULT_SET_UNACK,
+		BT_MESH_LIGHT_HSL_MSG_LEN_DEFAULT,
+		default_set_unack_handle,
+	},
+	{
+		BT_MESH_LIGHT_HSL_OP_RANGE_SET,
+		BT_MESH_LIGHT_HSL_MSG_LEN_RANGE_SET,
+		range_set_handle,
+	},
+	{
+		BT_MESH_LIGHT_HSL_OP_RANGE_SET_UNACK,
+		BT_MESH_LIGHT_HSL_MSG_LEN_RANGE_SET,
+		range_set_unack_handle,
+	},
+	BT_MESH_MODEL_OP_END,
+};
+
+static int hsl_srv_pub_update(struct bt_mesh_model *mod)
+{
+	struct bt_mesh_light_hsl_srv *srv = mod->user_data;
+	struct bt_mesh_light_hsl_status status;
+
+	hsl_get(srv, &status);
+	hsl_status_encode(&srv->buf, BT_MESH_LIGHT_HSL_OP_STATUS, &status);
+
+	return 0;
+}
+
+static int bt_mesh_light_hsl_srv_init(struct bt_mesh_model *model)
+{
+	struct bt_mesh_light_hsl_srv *srv = model->user_data;
+
+	srv->model = model;
+	srv->pub.update = hsl_srv_pub_update;
+	srv->pub.msg = &srv->buf;
+	srv->hue.hsl = srv;
+	srv->sat.hsl = srv;
+	srv->pub_pending = true;
+	net_buf_simple_init_with_data(&srv->buf, srv->pub_data,
+				      ARRAY_SIZE(srv->pub_data));
+
+	if (IS_ENABLED(CONFIG_BT_MESH_MODEL_EXTENSIONS)) {
+		/* Model extensions:
+		 * To simplify the model extension tree, we're flipping the
+		 * relationship between the Light HSL server and the Light HSL
+		 * setup server. In the specification, the Light HSL setup
+		 * server extends the time server, which is the opposite of
+		 * what we're doing here. This makes no difference for the mesh
+		 * stack, but it makes it a lot easier to extend this model, as
+		 * we won't have to support multiple extenders.
+		 */
+		bt_mesh_model_extend(model, srv->lightness.lightness_model);
+		bt_mesh_model_extend(
+			model, bt_mesh_model_find(
+				       bt_mesh_model_elem(model),
+				       BT_MESH_MODEL_ID_LIGHT_HSL_SETUP_SRV));
+	}
+
+	return 0;
+}
+
+static int bt_mesh_light_hsl_srv_start(struct bt_mesh_model *mod)
+{
+	struct bt_mesh_light_hsl_srv *srv = mod->user_data;
+	struct bt_mesh_model_transition transition;
+	struct bt_mesh_light_hue hue = { .transition = &transition };
+	struct bt_mesh_light_sat sat = { .transition = &transition };
+	uint16_t lightness =
+		(srv->lightness.ponoff.on_power_up == BT_MESH_ON_POWER_UP_OFF) ?
+			      0 :
```

Comment:
Lightness is a little different, as it will only use the default value on `ON_POWER_UP_ON`, while the saturation and hue have no natural "off" state. I did it your suggested way first, but this failed PTS tests.

Reply:
That's strange. If the bulb lost configuration or just from the shop it should light with some default value after power on.
I see the lightness model has a bit different behavior
```
	case BT_MESH_ON_POWER_UP_ON:
		set.lvl = (srv->default_light ? srv->default_light : srv->last);
		break;
```

Reply:
Yes, which will run first. This calls `lightness_srv_change_lvl`, which in turn sets `srv->last` to the chosen value. This function doesn't set the value, it just reads it out from the lightness server. This is super convoluted though, I'll try changing it, so it's readable.